File: Signal/test/Models/OWSDeviceProvisioningURLParserTest.m
//
//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
//

#import "OWSDeviceProvisioningURLParser.h"
#import "SignalBaseTest.h"
#import <SignalServiceKit/NSData+OWS.h>

@interface OWSDeviceProvisioningURLParserTest : SignalBaseTest

@end

@implementation OWSDeviceProvisioningURLParserTest

- (void)testValid
{
    OWSDeviceProvisioningURLParser *parser = [[OWSDeviceProvisioningURLParser alloc] initWithProvisioningURL:@""];
    XCTAssertFalse(parser.isValid);

    parser = [[OWSDeviceProvisioningURLParser alloc] initWithProvisioningURL:@"ts:/?uuid=MTIz"];
    XCTAssertFalse(parser.isValid);

    parser = [[OWSDeviceProvisioningURLParser alloc] initWithProvisioningURL:@"ts:/?pub_key=MTIz"];
    XCTAssertFalse(parser.isValid);

    parser = [[OWSDeviceProvisioningURLParser alloc] initWithProvisioningURL:@"ts:/uuid=asd&pub_key=MTIz"];
    XCTAssertFalse(parser.isValid);

    parser = [[OWSDeviceProvisioningURLParser alloc] initWithProvisioningURL:@"ts:/?uuid=asd&pub_key=MTIz"];
    XCTAssert(parser.isValid);
}

- (void)testPublicKey
{
    OWSDeviceProvisioningURLParser *parser =
        [[OWSDeviceProvisioningURLParser alloc] initWithProvisioningURL:@"ts:/?uuid=asd&pub_key=MTIz"];

    XCTAssertEqualObjects(@"MTIz", [parser.publicKey base64EncodedString]);
}

- (void)testEphemeralDeviceId
{
    OWSDeviceProvisioningURLParser *parser =
        [[OWSDeviceProvisioningURLParser alloc] initWithProvisioningURL:@"ts:/?uuid=asd&pub_key=MTIz"];

    XCTAssertEqualObjects(@"asd", parser.ephemeralDeviceId);
}

@end

File: Signal/test/OWSDeviceProvisioningURLParserTest.swift
//
// Copyright 2022 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import XCTest

class OWSDeviceProvisioningURLParserTest: XCTestCase {
    func testValid() {
        func isValid(_ provisioningURL: String) -> Bool {
            OWSDeviceProvisioningURLParser(provisioningURL: provisioningURL).isValid
        }

        XCTAssertFalse(isValid(""))
        XCTAssertFalse(isValid("ts:/?uuid=MTIz"))
        XCTAssertFalse(isValid("ts:/?pub_key=MTIz"))
        XCTAssertFalse(isValid("ts:/uuid=asd&pub_key=MTIz"))

        XCTAssertTrue(isValid("ts:/?uuid=asd&pub_key=MTIz"))
    }

    func testPublicKey() {
        let parser = OWSDeviceProvisioningURLParser(provisioningURL: "ts:/?uuid=asd&pub_key=MTIz")

        XCTAssertEqual(parser.publicKey?.base64EncodedString(), "MTIz")
    }

    func testEphemeralDeviceId() {
        let parser = OWSDeviceProvisioningURLParser(provisioningURL: "ts:/?uuid=asd&pub_key=MTIz")

        XCTAssertEqual(parser.ephemeralDeviceId, "asd")
    }
}
